src/qr.jl

```diff
@@ -0,0 +1,22 @@
+_stabilize_type(::Type{T}) where {T} = typeof((one(T)*zero(T) + zero(T))/one(T))
```

Comment:
We use this nice trick elsewhere too (det, solve, eigen, lu, inv).  How about tweaking the function to be called "factorization_eltype" - the expected eltype of the result of factorizing a matrix or solving a linear system:

```julia
factorization_eltype(a::AbstractMatrix{T}) where {T} = typeof((one(T)*zero(T) + zero(T))/one(T))
```

Reply:
Ah, that is a good idea. Maybe this deserves to go into Base because of issues like #23268, which can be solved by overloading `factorization_eltype`.

Reply:
Having this as a reusable function is a great idea. But it needs a good name.
`factorization_eltype` is not perfect since this function is also useful outside of factorizations (e.g. `det`). Also I would prefer to not bake `T = eltype(...)` into the function.
 What do you think of the following:
```Julia 
arithmetic_stabilization(T) = typeof((one(T)*zero(T) + zero(T))/one(T))
```

Reply:
How about `arithmetic_closure` - https://en.wikipedia.org/wiki/Closure_(mathematics) ("f that operation on members of the set always produces a member of the same set")

Reply:
In theory users might make a system that is not closed...

Reply:
We could make the function more "clever":
```Julia
arithmetic_closure1(T) = typeof((one(T)*zero(T) + zero(T))/one(T))
function arithmetic_closure(T)
    S = arithmetic_closure1(T)
    T == S ? S : arithmetic_closure(S)
end
```
or argue that its the users responsibility to overload it.

Reply:
Also if a user has such a system, she might want to redesign it anyway since it will cause subtle type stability issues in a lot of places?

Reply:
I'd like to keep mention of "stability" out of such a function name as this would be emphasizing an implementation detail rather than high level meaning of the function. This function is more about inferring the output type of certain types of high level operations, like matrix factorizations and other things.

I think the main question is how we'd like to group the high level operations.

For example, `det` and matrix multiplication do not use `/`, so maybe in these cases the output type should be inferred as `typeof(one(T)*zero(T) + zero(T))`, and we'd have a different name for this.

On that basis there could be two quite different ways to name this function:
* Operationally - closure under `*,+` vs closure under `*,+,/`, vs closure under `+` only, etc
* By category

On the whole, defining it operationally might be easiest and most well defined.

Reply:
I'm not so worried by the fact that the user might define an inconsistent and non-closed arithmetic. This seems relatively non-useful, and I can't imagine any useful way of automatically doing the right thing.  Providing a function which can be overridden for such crazy types would be perfectly fine by me ;-)

Reply:
It is true, that `det` mathematically makes sense in cases, where `/` is not defined. However efficient implementations often require `/`. So I think `det` should still use the above function. This is also the behavior in Base:
```Julia
julia> det([1 2; 3 4])
-2.0
```

Reply:
Okay maybe it is a weak argument that usage of `/` in the implementation affects the return type.

Reply:
We could do 
```Julia
closure(T) = T
function closure(T, op, ops...)
    t = one(T)
    S = typeof(op(t,t))
    closure(S, ops...)
end
```

Reply:
`det` is a tricky one.  It seems that it really needs to be implemented via factorization to avoid inefficiency and numerical problems, and it's prone to overflow even in stable implementations (hence `logdet`, I recently learned). So I think `Base` makes a reasonable and more or less practical choice here which we should follow for the moment.

The `closure` function is cute. I prefer the version you committed for now though. The term `closure` might be a bit problematic due to the conventional meaning being completely different in mathematics and programming...

Reply:
I agree with all you said. So this PR can be merged?